### DISCUSSION TITLE
Proposal for v4: Postgres RLS feature via Marten

### DISCUSSION THREAD

**Discussion by mysticmind** [JasperFx/marten#2111] [Ideas]
I am working out some nuts and bolts for Postgres row level security (RLS) feature via Marten. This will allow you to enforce RLS policies such as restrict querying by tenant or other data points.

Observe the following as below steps will be required:

- Enable RLS at the table level
- Define a Postgres policy using `using` or `with check` condition which uses a session variable
- Typical implementation is to call Postgres `set_config(setting_name, setting_val, is_local)` to set a session variable and use and use the value by getting it via `current_setting(setting_name)` in the policy condition.

In terms of Marten implementation:
1. We will need have document store schema settings to enable RLS at a table/document level
2. Define one or more postgres policies via document store schema
3. Have the ability to set a session variable using set_config on the spawned session or a specific transaction as a first step before running any other operations.

1 and 2 are more to do with schema management. (3) requires a pre-hook into Marten session to ensure that the session variables get set before any other operations.

**Comment by mdissel** [comment]
Referencing a closed issue, https://github.com/JasperFx/marten/issues/1352
